src/listbox/Listbox.ts

```diff
@@ -0,0 +1,264 @@
+				onActiveIndexChange && onActiveIndexChange(newIndex, key);
+				break;
+			case Keys.Up:
+				event.preventDefault();
+				newIndex = (activeIndex - 1 + optionData.length) % optionData.length;
+				onActiveIndexChange && onActiveIndexChange(newIndex, key);
+				break;
+			case Keys.Home:
+				onActiveIndexChange && onActiveIndexChange(0, key);
+				break;
+			case Keys.End:
+				onActiveIndexChange && onActiveIndexChange(optionData.length - 1, key);
+				break;
+		}
+	}
+
+	protected animateScroll(element: HTMLElement, scrollValue: number) {
+		element.scrollTop = scrollValue;
+	}
+
+	@diffProperty('activeIndex', auto)
+	protected calculateScroll(previousProperties: ListboxProperties, { activeIndex = 0 }: ListboxProperties) {
+		const scrollOffset = this.meta(Dimensions).get('root').scroll.top;
+		const menuHeight = this.meta(Dimensions).get('root').offset.height;
+		const optionOffset = this.meta(Dimensions).get(this._getOptionId(activeIndex)).offset;
+
+		if (optionOffset.top - scrollOffset < 0) {
+			this._scroll = optionOffset.top;
+			this.invalidate();
+		}
+
+		else if ((optionOffset.top + optionOffset.height) > (scrollOffset + menuHeight)) {
+			this._scroll = optionOffset.top + optionOffset.height - menuHeight;
+			this.invalidate();
+		}
+	}
+
+	protected getModifierClasses() {
+		const { visualFocus } = this.properties;
+		return [
+			visualFocus ? css.focused : null
+		];
+	}
+
+	protected getOptionClasses(active: boolean, disabled: boolean, selected: boolean) {
+		return [
+			css.option,
+			active ? css.activeOption : null,
+			disabled ? css.disabledOption : null,
+			selected ? css.selectedOption : null
+		];
+	}
+
+	protected onElementCreated(element: HTMLElement, key: string) {
+		if (key === 'root' && typeof this._scroll === 'number') {
+			this.animateScroll(element, this._scroll);
+			this._scroll = undefined;
+		}
+	}
+
+	protected onElementUpdated(element: HTMLElement, key: string) {
+		if (key === 'root' && typeof this._scroll === 'number') {
+			this.animateScroll(element, this._scroll);
+			this._scroll = undefined;
+		}
+	}
+
+	protected renderOptionLabel(option: any, index: number): DNode {
+		const { getOptionLabel } = this.properties;
+		return getOptionLabel ? getOptionLabel(option, index) : `${option}`;
+	}
+
+	protected renderOption(option: any, index: number): DNode {
+		const {
+			activeIndex = 0,
+			key = '',
+			getOptionDisabled = (option: any, index: number) => false,
+			getOptionSelected = (option: any, index: number) => false,
+			theme,
+			onActiveIndexChange,
+			onOptionSelect
+		} = this.properties;
+
+		const disabled = getOptionDisabled(option, index);
+		const selected = getOptionSelected(option, index);
+
+		return v('div', { key: this._getOptionId(index) }, [
+			w(ListboxOption, {
+				active: activeIndex === index,
+				classes: this.getOptionClasses(activeIndex === index, disabled, selected),
+				disabled,
+				label: this.renderOptionLabel(option, index),
```

Comment:
Should I break out `id` as well? I originally only pulled out `disabled` and `selected` because they're also used to get the state classes whereas `label`/`id` are only used once.